compiler/dcalc/ast.mli

```diff
@@ -131,6 +131,41 @@ type enum_ctx = (EnumConstructor.t * typ Pos.marked) list EnumMap.t
 
 type decl_ctx = { ctx_enums : enum_ctx; ctx_structs : struct_ctx }
 
+type binder = (expr, expr Pos.marked) Bindlib.binder
+
+(** This kind annotation signals that the let-binding respects a structural invariant. These
+    invariants concern the shape of the expression in the let-binding, and are documented below. *)
+type scope_let_kind =
+  | DestructuringInputStruct  (** [let x = input.field]*)
+  | ScopeVarDefinition  (** [let x = error_on_empty e]*)
+  | SubScopeVarDefinition  (** [let s.x = fun _ -> e] *)
+  | CallingSubScope  (** [let result = s ({ x = s.x; y = s.x; ...}) ]*)
+  | DestructuringSubScopeResults  (** [let s.x = result.x ]**)
+  | Assertion  (** [let _ = assert e]*)
```

Comment:
thanks for the comments! :)

Reply:
after reading the scope to dcalc transformation, it might be a good idea here to put

[let x = e, where e = error_on_empty e' for some e']

to emphasis the fact that structural invariant are not kept by the ocaml type system.

Reply:
or we could use gadts, but that's more complicated